### PR TITLE
Commit header files for L2/L3 multicast

### PR DESCRIPTION
**Pull request by sonicyao1985**

### CODE REVIEW

File: inc/sai.h [hunk 1]
```diff
@@ -77,7 +79,7 @@ typedef enum _sai_api_t
     SAI_API_VIRTUAL_ROUTER   =  5, /**< sai_virtual_router_api_t */
     SAI_API_ROUTE            =  6, /**< sai_route_api_t */
     SAI_API_NEXT_HOP         =  7, /**< sai_next_hop_api_t */
-    SAI_API_NEXT_HOP_GROUP   =  8, /**< sai_next_hop_group_api_t */
+    SAI_API_GROUP            =  8, /**< sai_group_api_t */
```
Comment (atitjain) [inline inc/sai.h]: A better descriptive name would be better. Say ENDPOINT_GROUP or EGRESS_GROUP etc.

File: inc/saigroup.h [hunk 2]
```diff
@@ -0,0 +1,826 @@
...
+
+/**
+ * @brief Next hop group type
+ */
+typedef enum _sai_next_hop_group_type_t
+{
+    /** Next hop group is ECMP */
+    SAI_NEXT_HOP_GROUP_TYPE_ECMP,
+
+    /* Other types of next hop group to be defined in the future, e.g., WCMP */
```
Comment (atitjain) [inline inc/saigroup.h]: would there be more types as you have made group as generic object for any collection of egresses.

File: inc/saigroup.h [hunk 3]
```diff
@@ -0,0 +1,826 @@
...
+    sai_create_next_hop_group_fn               create_next_hop_group;
+    sai_remove_next_hop_group_fn               remove_next_hop_group;
+    sai_set_next_hop_group_attribute_fn        set_next_hop_group_attribute;
+    sai_get_next_hop_group_attribute_fn        get_next_hop_group_attribute;
+    sai_create_next_hop_group_member_fn        create_next_hop_group_member;
+    sai_remove_next_hop_group_member_fn        remove_next_hop_group_member;
+    sai_set_next_hop_group_member_attribute_fn set_next_hop_group_member_attribute;
+    sai_get_next_hop_group_member_attribute_fn get_next_hop_group_member_attribute;
+
+    sai_create_rpf_group_fn                    create_rpf_group;
```
Comment (atitjain) [inline inc/saigroup.h]: Other than the same file, we dont seem to reuse any of the existing group things. If we need to add all these new APIs why not create a new file and objects like RPF group **mc etc as part of that.

File: inc/saiipmc.h [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+     * @type sai_object_id_t
+     * @objects SAI_OBJECT_TYPE_IPMC_GROUP
+     * @flags MANDATORY_ON_CREATE | CREATE_AND_SET
+     */
+    SAI_IPMC_ENTRY_ATTR_OUTPUT_GROUP_ID,
+
+    /**
+     * @brief IPMC entry RPF interface group id
+     *
+     * If not set or the group has no member, RPF checking will be disabled
```
Comment (atitjain) [inline inc/saiipmc.h]: if not set - do you mean set to null?

File: inc/sai.h (hunk 1 above)
Comment (sonicyao1985) [inline inc/sai.h]: not good, RPF interface group is not either ENDPOINT or EGRESS. It is ingress. Better keep it this way. : )

File: inc/saifdb.h [hunk 5]
```diff
@@ -95,17 +95,18 @@ typedef enum _sai_fdb_entry_attr_t
...
-     * The port id here can refer to a generic port object such as SAI port object id,
+     * The output id here can refer to a generic port object such as SAI port object id,
      * SAI LAG object id and etc. or to a tunnel next hop object in case the entry is
-     * l2 tunnel
+     * l2 tunnel. It can also refer to a L2MC group object. In case of empty group, packets 
+     * will be discarded
      *
      * @type sai_object_id_t
-     * @objects SAI_OBJECT_TYPE_PORT, SAI_OBJECT_TYPE_LAG, SAI_OBJECT_TYPE_TUNNEL
+     * @objects SAI_OBJECT_TYPE_PORT, SAI_OBJECT_TYPE_LAG, SAI_OBJECT_TYPE_TUNNEL, SAI_OBJECT_TYPE_L2MC_GROUP
```
Comment (sonicyao1985) [inline inc/saifdb.h]: nope. generic group is rejected. so now we have several group objects as l2mc_group/ipmc_group/rpf_group

File: inc/saiipmc.h (hunk 4 above)
Comment (sonicyao1985) [inline inc/saiipmc.h]: yes.

File: inc/saigroup.h (hunk 2 above)
Comment (sonicyao1985) [inline inc/saigroup.h]: hi aititjain, the concept of generic group is rejected by Microsoft & MLNX in one of the weekly meetings.

File: inc/saigroup.h (hunk 3 above)
Comment (sonicyao1985) [inline inc/saigroup.h]: ok. I will create new files for different groups.

File: inc/saifdb.h [hunk 6]
```diff
@@ -95,17 +95,18 @@ typedef enum _sai_fdb_entry_attr_t
...
      *
-     * The port id here can refer to a generic port object such as SAI port object id,
+     * The output id here can refer to a generic port object such as SAI port object id,
      * SAI LAG object id and etc. or to a tunnel next hop object in case the entry is
-     * l2 tunnel
+     * l2 tunnel. It can also refer to a L2MC group object. In case of empty group, packets 
+     * will be discarded
      *
      * @type sai_object_id_t
-     * @objects SAI_OBJECT_TYPE_PORT, SAI_OBJECT_TYPE_LAG, SAI_OBJECT_TYPE_TUNNEL
```
Comment (sonicyao1985) [inline inc/saifdb.h]: I'm not sure how mlnx will implement the header files of the bridge port proposal. If this part is not covered, I will take care of it later.

File: inc/sai.h (hunk 1 above)
Comment (atitjain) [inline inc/sai.h]: but SAI_API_GROUP is not as descriptive. Sounds like Group of APIs. May be add RPF Source group .

**Comment by lguohan** [comment]
@sonicyao1985 , can you help to resolve the conflict and also address the review comments?

File: inc/saiipmc.h [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    SAI_IPMC_ENTRY_TYPE_XG,
+
+} sai_ipmc_entry_type_t;
+
+/**
+ * @brief IPMC entry key
+ */
+typedef struct _sai_ipmc_entry_t
+{
+    /** VRF ID */
```
Comment (lguohan) [inline inc/saiipmc.h]: need switch id in the entry

**Comment by sonicyao1985** [comment]
will make new PR

**Pull request closed without merging**